query-planner-js/src/buildPlan.ts

```diff
@@ -1817,35 +1829,86 @@ class GroupPath {
-      return this.responsePath;
-    }
+  private updatedResponsePath(element: OperationElement): ResponsePath {
+    switch (element.kind){
+      case 'FragmentElement':
+        return this.responsePath;
+      case 'Field':
+        let newPath = this.responsePath;
+        if (this.possibleTypesAfterLastField.length !== this.possibleTypes.length) {
+          const conditions = `|[${this.possibleTypes.join(',')}]`;
+          const previousLastElement = newPath[newPath.length -1] as string || '';
 
-    let type = element.definition.type!;
-    const newPath = this.responsePath.concat(element.responseName());
-    while (!isNamedType(type)) {
-      if (isListType(type)) {
-        newPath.push('@');
-      }
-      type = type.ofType;
+          if (previousLastElement.startsWith('|[')) {
+            newPath = [...newPath.slice(0, -1), conditions];
+          } else {
+            newPath = [...newPath.slice(0, -1), `${previousLastElement}${conditions}`];
+          }
+        }
+        let type = element.definition.type!;
+        if (newPath.length === 0 && this.typeConditionedFetching) {
+          newPath = newPath.concat('');
+        }
+        newPath = newPath.concat(`${element.responseName()}`);
+        while (!isNamedType(type)) {
+          if (isListType(type)) {
+            newPath.push('@');
+          }
+          type = type.ofType;
+        }
+        return newPath;
     }
-    return newPath;
   }
 
   add(element: OperationElement): GroupPath {
+    const responsePath = this.updatedResponsePath(element);
+    const newPossibleTypes = this.computeNewPossibleTypes(element);
     return new GroupPath(
       this.fullPath.concat(element),
       this.pathInGroup.concat(element),
-      this.updatedResponsePath(element),
+      responsePath,
+      this.typeConditionedFetching,
+      newPossibleTypes,
+      element.kind === 'Field'? newPossibleTypes: this.possibleTypesAfterLastField
     );
   }
 
   toString() {
-    return `[${this.fullPath}]:[${this.pathInGroup}]`;
+    return this.inResponse().join('.');
+  }
+
+  computeNewPossibleTypes(element: OperationElement): ObjectType[] {
+    if (!this.typeConditionedFetching) {
+      return [];
+    }
+    switch (element.kind){
+      case 'FragmentElement':
+        if (!element.typeCondition) {
+          return this.possibleTypes;
+        }
+        const elementPossibleTypes = possibleRuntimeTypes(element.typeCondition);
+        return this.possibleTypes.filter((pt) => elementPossibleTypes.some((ept) => ept.name === pt.name));
+      case 'Field':
+        return this.advanceFieldType(element);
+    }
+  }
+
+
+  advanceFieldType(element: Field): ObjectType[] {
+    if (!isCompositeType(element.baseType())) {
+      return [];
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const res = Array.from(new Set(this.possibleTypes.map((pt) => possibleRuntimeTypes(baseType(pt.field(element.name)!.type!) as CompositeType)).flat()));
```

Comment:
From maintenance point of view this line is super complex (map => flatten => to set => to array). Wondering whether formatting it across multiple lines would make it clearer (guessing es-lint or some other auto formatter would complain anyway....)

```typescript
const res = Array.from(
    new Set(
        this.possibleTypes.map(
            (pt) => possibleRuntimeTypes(
                baseType(pt.field(element.name)!.type!) as CompositeType
            )
        ).flat()
    )
)
```

Reply:
fixed in eb0b6e44